src/wallet/index.ts

```diff
@@ -190,4 +200,111 @@ class Wallet {
+  const tx = JSON.parse(txJSON)
+  if (tx.TxnSignature || tx.Signers) {
+    throw new ValidationError(
+      'txJSON must not contain "TxnSignature" or "Signers" properties',
+    )
+  }
+
+  const txToSignAndEncode = { ...tx }
+
+  txToSignAndEncode.SigningPubKey = options.signAs ? '' : publicKey
+
+  if (options.signAs) {
+    const signer = {
+      Account: options.signAs,
+      SigningPubKey: publicKey,
+      TxnSignature: computeSignature(
+        txToSignAndEncode,
+        privateKey,
+        options.signAs,
+      ),
+    }
+    txToSignAndEncode.Signers = [{ Signer: signer }]
+  } else {
+    txToSignAndEncode.TxnSignature = computeSignature(
+      txToSignAndEncode,
+      privateKey,
+    )
+  }
+  const serialized = encode(txToSignAndEncode)
+  checkTxSerialization(serialized, tx)
+  return serialized
+}
+
+/**
+ *  Decode a serialized transaction, remove the fields that are added during the signing process,
+ *  and verify that it matches the transaction prior to signing. This gives the user a sanity check
+ *  to ensure that what they try to encode matches the message that will be recieved by rippled.
+ *
+ * @param serialized - A signed and serialized transaction.
+ * @param tx - The transaction prior to signing.
+ *
+ * @returns This method does not return a value
+ * @throws a ValidationError if the transaction does not have a TxnSignature/Signers property, or an AssertionError if
+ * the serialized Transaction desn't match the original transaction.
+ */
+function checkTxSerialization(serialized: string, tx: Transaction): void {
+  // Decode the serialized transaction:
+  const decoded = decode(serialized)
+
+  // ...And ensure it is equal to the original tx, except:
+  // - It must have a TxnSignature or Signers (multisign).
+  if (!decoded.TxnSignature && !decoded.Signers) {
+    throw new ValidationError(
+      'Serialized transaction must have a TxnSignature or Signers property',
+    )
+  }
+  // - We know that the original tx did not have TxnSignature, so we should delete it:
+  delete decoded.TxnSignature
+  // - We know that the original tx did not have Signers, so if it exists, we should delete it:
+  delete decoded.Signers
+
+  // - If SigningPubKey was not in the original tx, then we should delete it.
+  //   But if it was in the original tx, then we should ensure that it has not been changed.
+  if (!tx.SigningPubKey) {
+    delete decoded.SigningPubKey
+  }
+
+  // - Memos have exclusively hex data which should ignore case.
+  //   Since decode goes to upper case, we set all tx memos to be uppercase for the comparison.
+  tx.Memos?.map((memo) => {
+    if (memo.Memo.MemoData) {
+      memo.Memo.MemoData = memo.Memo.MemoData.toUpperCase()
+    }
+
+    if (memo.Memo.MemoType) {
+      memo.Memo.MemoType = memo.Memo.MemoType.toUpperCase()
+    }
+
+    if (memo.Memo.MemoFormat) {
+      memo.Memo.MemoFormat = memo.Memo.MemoFormat.toUpperCase()
+    }
+
+    return memo
+  })
+
+  assert.deepEqual(
```

Comment:
Can we put this in a try/catch and throw a different error than an `AssertionError`?